diff --git a/src/connection.js b/src/connection.js
--- a/src/connection.js
+++ b/src/connection.js
@@ -564,6 +564,12 @@
         return fetchArrayTypes()
       }
 
+      if (initial.reserve) {
+        initial = null
+        options.shared.retries = retries = 0
+        return onopen(connection)
+      }
+
       initial && !initial.reserve && execute(initial)
       options.shared.retries = retries = 0
       initial = null
diff --git a/cjs/src/connection.js b/cjs/src/connection.js
--- a/cjs/src/connection.js
+++ b/cjs/src/connection.js
@@ -564,6 +564,12 @@
         return fetchArrayTypes()
       }
 
+      if (initial.reserve) {
+        initial = null
+        options.shared.retries = retries = 0
+        return onopen(connection)
+      }
+
       initial && !initial.reserve && execute(initial)
       options.shared.retries = retries = 0
       initial = null
diff --git a/cf/src/connection.js b/cf/src/connection.js
--- a/cf/src/connection.js
+++ b/cf/src/connection.js
@@ -564,6 +564,12 @@
         return fetchArrayTypes()
       }
 
+      if (initial.reserve) {
+        initial = null
+        options.shared.retries = retries = 0
+        return onopen(connection)
+      }
+
       initial && !initial.reserve && execute(initial)
       options.shared.retries = retries = 0
       initial = null
diff --git a/cf/polyfills.js b/cf/polyfills.js
--- a/cf/polyfills.js
+++ b/cf/polyfills.js
@@ -211,7 +211,8 @@
   }
 
   function error(err) {
-    tcp.emit('error', err)
+    if (tcp.listenerCount('error') > 0)
+      tcp.emit('error', err)
     tcp.emit('close')
   }
 }
